tests/unittests/RecommendationSystemTest.cpp

```diff
@@ -0,0 +1,678 @@
+          ElemKind::Int32ITy, {miniBatch}, "SL" + std::to_string(i), false);
+    }
+
+    auto *denseData = mod_.createPlaceholder(
+        ElemKind::FloatTy, {miniBatch, denseDim}, "denseData",
+        false); // Dense dim can be anything
+
+    // Generate the network.
+    createSimpleRecSysGraph(mod_, bindings_, F_, sparseLengths, denseData,
+                            tableSizes, embeddingDim, "main", quantizeSLWS,
+                            quantizeFC);
+    SaveNode *result1 = llvm::cast<SaveNode>(F_->getNodeByName("save"));
+    result = result1->getPlaceholder();
+
+    if (convertToFP16) {
+      KindSet keepOriginalPrecisionForNodes;
+      keepOriginalPrecisionForNodes.insert(
+          Kinded::Kind::FusedRowwiseQuantizedSparseLengthsWeightedSumNodeKind);
+      keepOriginalPrecisionForNodes.insert(
+          Kinded::Kind::RowwiseQuantizedFullyConnectedNodeKind);
+      TypeAToTypeBFunctionConverter converter(*F_, ElemKind::FloatTy,
+                                              ElemKind::Float16Ty,
+                                              &keepOriginalPrecisionForNodes);
+      converter.convert();
+      ::optimize(F_, glow::CompilationMode::Infer);
+    }
+
+    // Compile.
+    EE_.compile(CompilationMode::Infer, F_);
+
+    // Run graph
+    EE_.run(bindings_);
+
+    // Get result and verify.
+    auto *resultTensor = bindings_.get(result);
+    auto resultHandle = resultTensor->getHandle();
+
+    EXPECT_EQ(resultTensor->size(), miniBatch);
+    for (int i = 0, e = resultHandle.size(); i < e; ++i) {
+      EXPECT_LE(resultHandle.raw(i), 1.0);
+    }
+  }
+
+  /// Execute a graph of functions based on the given DAG.
+  void executeDAG(DAGNode *G, Module &mod, PlaceholderBindings &bindings,
+                  llvm::ArrayRef<Placeholder *> vars,
+                  llvm::ArrayRef<Tensor *> inputs) {
+    std::unordered_map<std::string, Function *> name2func;
+
+    for (auto *F : mod.getFunctions()) {
+      name2func[F->getName()] = F;
+    }
+
+    std::vector<DAGNode *> exeList;
+    int endPt = 0;
+    int curPt = 0;
+    // The first node is always the dummy node.
+    exeList.push_back(G);
+    endPt++;
+    while (curPt < endPt) {
+      DAGNode *dag = exeList.at(curPt);
+      // The root in a G is always a dummy function.
+      if (curPt > 0) {
+        ExecutionEngine EE{EE_.getBackend()->getBackendKind()};
+        Function *func = name2func[dag->name];
+        EE.compile(CompilationMode::Infer, func);
+        updateInputPlaceholders(bindings, vars, inputs);
+        EE.run(bindings);
+      }
+      for (int i = 0, e = dag->children.size(); i < e; i++) {
+        exeList.push_back(dag->children.at(i));
+        endPt++;
+      }
+      curPt++;
+    }
+  }
+
+  /// Create partitions to run and compare results.
+  void runPartitionedGraph(size_t numDevices, size_t memSize,
+                           Tensor referenceResult,
+                           PlaceholderBindings &bindings) {
```

Comment:
Can we apply HostManager to do Runtime End2End test directly?

Reply:
Yes we should. Right now its blocked on #2874 and some additional plumbing so we can set device sizes and force the partitioning, I've created #2897 for this.